third_party/amd/lib/TritonAMDGPUTransforms/CanonicalizePointers.cpp

```diff
@@ -737,6 +739,47 @@ LogicalResult PointerCanonicalizer::rewriteCondBranchOp(
   return success();
 }
 
+LogicalResult PointerCanonicalizer::rewriteSelectOp(arith::SelectOp selectOp,
+                                                    Location curLoc,
+                                                    OpOperand *curOperand,
+                                                    Value &nextPtr) {
+  const auto &operands = selectOp->getOpOperands();
+  bool otherIsFalse = &operands[1] == curOperand;
+  assert(otherIsFalse || &operands[2] == curOperand);
+
+  OpOperand *otherOperand = &operands[otherIsFalse ? 2 : 1];
+
+  // If we didn't traverse both operands, simply create the pointers
+  // for the operand we reached and save the fat pointer back. When (and if)
+  // the traversal meets the select again we can rewrite it properly
+  if (!pointers.contains(otherOperand->get())) {
+    FatPtr fatPtr = pointers[curOperand->get()];
+    Value tensorPtr = createTensorPointer(fatPtr, curLoc);
+    curOperand->set(tensorPtr);
+    pointers[curOperand->get()] = fatPtr;
```

Comment:
Isn't this just resetting the same value from line 756?

Reply:
Yes, but on a different operand. Original situation is:
```
%ptr3 =  select %ptr1, %ptr2 // pointers[ptr1] = fatPtr1
```
When I rewrite that I do:
```
%s = splat fatPtr1.basePtr
%p = addptr %s, fatPtr1.offset
%ptr3 = select %p, %ptr2
```
Now I have to tell what is the fat pointer of `p` (this is line 759). So after line 759, the situation becomes:
```
%ptr3 = select %p, %ptr2 //fatPtr[p] = fatPtr1
```
So if I don't traverse `ptr2`, the propagation stops and the code is in a good state. If I traverse it the code can be rewritten with the `select base` + `select offset`

Reply:
I'm still confused about the logic here--afaict, you are just creating the splatted base `%s` but not adding the offset so no `%p` here? 

Do we really need to update the existing select op in place? that's typically a footgun--we already had issues that gotten fixed before due to some inplace modification. Is there a better way to handle this?

Reply:
Yeah you were right, that was not correct. So now I changed the `createTensorPointer` so that it splat thebase and add the offset. 

Also, instead of changing in place, I am calling `materializeFatPointer`, which is creating the tensor pointer and cloning the operation (with the right mapping).